src/hotspot/share/code/icBuffer.cpp

```diff
@@ -247,26 +248,42 @@ void* InlineCacheBuffer::cached_value_for(CompiledIC *ic) {
 // Free CompiledICHolder*s that are no longer in use
 void InlineCacheBuffer::release_pending_icholders() {
   assert(SafepointSynchronize::is_at_safepoint(), "should only be called during a safepoint");
-  CompiledICHolder* holder = _pending_released;
+  CompiledICHolder* holder = Atomic::load(&_pending_released);
   _pending_released = nullptr;
+  int count = 0;
   while (holder != nullptr) {
     CompiledICHolder* next = holder->next();
     delete holder;
     holder = next;
-    _pending_count--;
+    count++;
   }
-  assert(_pending_count == 0, "wrong count");
+  assert(pending_icholder_count() == count, "wrong count");
+  Atomic::store(&_pending_count, 0);
 }
 
 // Enqueue this icholder for release during the next safepoint.  It's
-// not safe to free them until them since they might be visible to
+// not safe to free them until then since they might be visible to
 // another thread.
 void InlineCacheBuffer::queue_for_release(CompiledICHolder* icholder) {
-  MutexLocker mex(InlineCacheBuffer_lock, Mutex::_no_safepoint_check_flag);
-  icholder->set_next(_pending_released);
-  _pending_released = icholder;
-  _pending_count++;
+  assert(icholder->next() == nullptr, "multiple enqueue?");
+
+  CompiledICHolder* old = Atomic::load(&_pending_released);
+  for (;;) {
+    icholder->set_next(old);
```

Comment:
Could we do this only after the swap is successful?

Reply:
I think it could be done since nobody is actually looking at the list until a safepoint where it should be fixed up. What would be the benefit of doing this?

The current implementation is imo a fairly idiomatic/unsurprising way adding nodes to a linked list lock-free (in gc code, I did not look for examples in compiler code). I would probably expect any significiant deviation to be commented which seems unnecessary and just asking for the question "why is it done this way?".

The only thing I could think of would be that it would allow one to put the `next == nullptr` assert later, but that does not seem to be worth doing - the `CompiledICHolder`s are not shared, i.e. that multiple threads may add the same item (the previous locking code would have failed in this situation already). Executing the assert later will imo not significantly increase the possibility of failure (assuming that the far majority of additions are uncontended).

Reply:
It would avoid a unneeded set_next every time we loop on contention, but it's probably not worth changing just to optimize the contended case, and like you said the code is less surprising this way, so just leave it.